### PR TITLE
CASEFLOW-948 CAVC entry form: JMR/JMPR dynamic issue validation

### PR DESCRIPTION
**Pull request by yoomlam**
Resolves [CASEFLOW-948](https://vajira.max.gov/browse/CASEFLOW-948)

As a CAVC Litigation Support user, when I select/de-select issues, I want an informational banner noting all issues need to be selected for JMR Remands and at least one needs to be selected for JMPR, so that I do not have to hit submit before I see an error. 

### Description
For the CAVC entry form when JMR or JMPR is selected, show informational banner based on number of issues selected.

### Acceptance Criteria
- [x] Code compiles correctly

Scenario: Given I am a CAVC Litigation Support user entering a CAVC Remand
When I identify the remand type is JMR and I de-select one or more issues
- [x] Then an informational banner displays below the issues section with the following copy: "For a JMR you must select all issues to proceed"

When I identify the remand type is a JMPR **or MDR** and I de-select all issues
- [x] Then an informational banner displays below the issues section with the following copy: "For a JMPR you must select at least one issue to proceed" 

(The MDR requirement was added during a demo to Ann-Marie.)

### Testing Plan
1. Open a Rails console and create a ready for CAVC appeal, note the `stream_docket_number` so you can search for the appeal in the UI or the `uuid`.
```ruby
original_appeal = FactoryBot.create(:appeal, :dispatched, associated_judge: JudgeTeam.first.judge, associated_attorney: JudgeTeam.first.attorneys.first);
FactoryBot.create_list(:request_issue, 2, :rating, :with_rating_decision_issue, decision_review: original_appeal, veteran_participant_id: Veteran.first.participant_id, contested_issue_description: "issue_description", notes: "issue_notes");
original_appeal.uuid
```
2. Login as a CAVC lit support user 
3. Search for the dispatched appeal you just created and select it to go to Case Details: `http://localhost:3000/queue/appeals/<UUID>`
4. Click the "+ Add CAVC Remand" button in the "Post-Dispatch actions" section
5. Fill out the "Add a Court Remand" form
   * select JMR: Unselect an issue and notice the informational banner.
   * select JMPR: Unselect all issue and notice the informational banner.
   * select MDR: Unselect all issue and notice the informational banner.
6. Play around, stress test, and find bugs early

### User Facing Changes
 - [ ] Screenshots of UI changes

 BEFORE|AFTER
 ---|---
image | image
image | image

### Storybook Story
* [ ] ~Write a separate story (within the same file) for each discrete variation of the component~ This is not a sufficiently controlled component, so I can change the remand subtype or select issues.

### CODE REVIEW

File: client/app/queue/AddCavcRemandView.jsx [hunk 1]
```diff
@@ -102,15 +104,15 @@ const AddCavcRemandView = (props) => {
     label: decisionIssue.description
   }));
 
-  // determines which issues are currently selected
-  const selectedIssues = useMemo(() => {
+  // returns ids of issues that are currently selected
+  const selectedIssueIds = useMemo(() => {
```
Comment (yoomlam) [inline client/app/queue/AddCavcRemandView.jsx:108]: (Not relevant to this PR) Clarify the purpose of this function.

File: client/app/queue/AddCavcRemandView.jsx [hunk 2]
```diff
@@ -27,6 +27,7 @@ import { withRouter } from 'react-router';
 
 const radioLabelStyling = css({ marginTop: '2.5rem' });
 const buttonStyling = css({ paddingLeft: '0' });
+const bottomInfoStyling = css({ marginBottom: '4rem' });
```
Comment (yoomlam) [inline client/app/queue/AddCavcRemandView.jsx:30]: Adds more bottom margin than the default.
[Slack convo](https://dsva.slack.com/archives/CQCUMFA0H/p1614189059053200?thread_ts=1613758269.106800&cid=CQCUMFA0H)

File: client/app/queue/AddCavcRemandView.jsx [hunk 3]
```diff
@@ -120,13 +122,18 @@ const AddCavcRemandView = (props) => {
   // populate all of our checkboxes on initial render
   useEffect(() => selectAllIssues(), []);
 
+  const allIssuesSelected = useMemo(() => Object.values(issues).every((isChecked) => isChecked), [issues]);
+  const allIssuesUnselected = useMemo(() => Object.values(issues).every((isChecked) => !isChecked), [issues]);
+
```
Comment (yoomlam) [inline client/app/queue/AddCavcRemandView.jsx:127]: Could combine this into a single method, but this is more readable.

File: client/app/queue/AddCavcRemandView.jsx [hunk 4]
```diff
@@ -135,7 +142,7 @@ const AddCavcRemandView = (props) => {
   const validDocketNumber = () => (/^\d{2}[-‐−–—]\d{1,5}$/).exec(docketNumber);
   const validJudge = () => Boolean(judge);
   const validDecisionDate = () => Boolean(decisionDate) && validateDateNotInFuture(decisionDate);
-  const validDecisionIssues = () => selectedIssues && selectedIssues.length > 0;
+  const validDecisionIssues = () => selectedIssueIds?.length > 0;
```
Comment (yoomlam) [inline client/app/queue/AddCavcRemandView.jsx:145]: (Not relevant to this PR) Use [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) for cleaner code.

File: client/app/queue/AddCavcRemandView.jsx [hunk 5]
```diff
@@ -348,15 +365,17 @@ const AddCavcRemandView = (props) => {
       {representedField}
       {judgeField}
       {typeField}
-      {type === CAVC_DECISION_TYPES.remand && remandTypeField }
+      {remandType() && remandTypeField }
```
Comment (yoomlam) [inline client/app/queue/AddCavcRemandView.jsx:368]: Clean up: use existing method.

**Comment by araposo-tistatech** [comment]
Adding MDR to this as well

File: client/test/app/queue/__snapshots__/AddCavcRemandView.test.js.snap [hunk 6]
```diff
@@ -59,11 +59,11 @@ exports[`AddCavcRemandView renders correctly 1`] = `
         Array [
           Object {
             "description": "This is a description of the decision",
-            "id": 1,
+            "id": "1",
```
Comment (yoomlam) [inline client/test/app/queue/__snapshots__/AddCavcRemandView.test.js.snap:62]: Addresses "String expected" error when running Jest test.

File: client/test/data/appeals.js [hunk 7]
```diff
@@ -72,10 +72,10 @@ export const appealData = {
   issues: [],
   decisionIssues: [
     {
-      id: 1,
+      id: '1',
```
Comment (yoomlam) [inline client/test/data/appeals.js:75]: Addresses "String expected" error when running Jest test.

File: client/app/queue/AddCavcRemandView.jsx [hunk 8]
```diff
@@ -98,25 +101,28 @@ const AddCavcRemandView = (props) => {
   const filteredRemandTypes = subTypeOptions.filter((subTypeOption) => supportedRemandTypes[subTypeOption.value]);
 
   const issueOptions = () => decisionIssues.map((decisionIssue) => ({
-    id: decisionIssue.id,
+    id: decisionIssue.id.toString(),
```
Comment (yoomlam) [inline client/app/queue/AddCavcRemandView.jsx:104]: Addresses "String expected" error when running Jest test.

File: client/app/queue/AddCavcRemandView.jsx (hunk 3 above)
Comment (ajspotts) [reply client/app/queue/AddCavcRemandView.jsx:125]: agreed, also nice usage of `useMemo`

**Review by ajspotts — APPROVED** [review]
nice changes, was able to navigate between the different subtypes in combination with the various issue selections and the banners behaved properly. LGTM

**Pull request merged**